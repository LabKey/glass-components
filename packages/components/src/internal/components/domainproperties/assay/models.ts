/*
 * Copyright (c) 2019 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { List, Record as ImmutableRecord } from 'immutable';

import { getServerContext, Utils } from '@labkey/api';

import { DomainDesign, FieldErrors } from '../models';
import { AppURL } from '../../../url/AppURL';
import { getAppHomeFolderPath } from '../../../app/utils';
import { Container } from '../../base/models/Container';

// See ExpProtocol.Status in 'platform' repository.
export enum Status {
    Active = 'Active',
    Archived = 'Archived',
}

export interface ProtocolTransformScript {
    runOnEdit: boolean;
    runOnImport: boolean;
    scriptPath: string;
}

export class AssayProtocolModel extends ImmutableRecord({
    allowBackgroundUpload: false,
    allowEditableResults: false,
    allowQCStates: false,
    allowTransformationScript: false,
    allowPlateMetadata: false,
    autoCopyTargetContainer: undefined,
    autoCopyTargetContainerId: undefined,
    autoLinkCategory: undefined,
    availableDetectionMethods: undefined,
    availableMetadataInputFormats: undefined,
    availablePlateTemplates: undefined,
    backgroundUpload: false,
    description: undefined,
    domains: undefined,
    editableResults: false,
    editableRuns: false,
    exception: undefined,
    metadataInputFormatHelp: undefined,
    moduleTransformScripts: undefined,
    name: undefined,
    plateMetadata: undefined,
    protocolId: undefined,
    protocolParameters: undefined,
    protocolTransformScripts: undefined,
    providerName: undefined,
    saveScriptFiles: false,
    selectedDetectionMethod: undefined,
    selectedMetadataInputFormat: undefined,
    selectedPlateTemplate: undefined,
    status: Status.Active,
    qcEnabled: undefined,
    excludedContainerIds: undefined,
}) {
    declare allowBackgroundUpload: boolean;
    declare allowEditableResults: boolean;
    declare allowQCStates: boolean;
    declare allowTransformationScript: boolean;
    declare allowPlateMetadata: boolean;
    declare autoCopyTargetContainer: {};
    declare autoCopyTargetContainerId: string;
    declare autoLinkCategory: string;
    declare availableDetectionMethods: [];
    declare availableMetadataInputFormats: {};
    declare availablePlateTemplates: [];
    declare backgroundUpload: boolean;
    declare description: string;
    declare domains: List<DomainDesign>;
    declare editableResults: boolean;
    declare editableRuns: boolean;
    declare exception: string;
    declare metadataInputFormatHelp: any;
    declare moduleTransformScripts: List<string>;
    declare name: string;
    declare plateMetadata: boolean;
    declare protocolId: number;
    declare protocolParameters: any;
    declare protocolTransformScripts: List<ProtocolTransformScript>;
    declare providerName: string;
    declare saveScriptFiles: boolean;
    declare selectedDetectionMethod: string;
    declare selectedMetadataInputFormat: string;
    declare selectedPlateTemplate: string;
    declare status: Status;
    declare qcEnabled: boolean;
    declare excludedContainerIds: string[];

    static create(raw: any): AssayProtocolModel {
        let domains = raw.domains || List<DomainDesign>();
        if (raw.domains && Utils.isArray(raw.domains)) {
            domains = List<DomainDesign>(
                raw.domains.map(domain => {
                    return DomainDesign.create(domain);
                })
            );
        }

        if (raw.protocolTransformScripts && Utils.isArray(raw.protocolTransformScripts)) {
            raw.protocolTransformScripts = List(raw.protocolTransformScripts);
        }
        if (raw.moduleTransformScripts && Utils.isArray(raw.moduleTransformScripts)) {
            raw.moduleTransformScripts = List<string>(raw.moduleTransformScripts);
        }

        // if this is not an existing assay, clear the name property so the user must set it
        const name = !raw.protocolId ? undefined : raw.name;

        // Issue 38685: for new assays, pre-select some required assay properties
        const model = new AssayProtocolModel({ ...raw, name, domains });
        if (model.isNew()) {
            if (model.allowDetectionMethodSelection() && model.availableDetectionMethods.length > 0) {
                raw.selectedDetectionMethod = List<string>(model.availableDetectionMethods).get(0);
            }
            if (model.allowPlateTemplateSelection() && model.availablePlateTemplates.length > 0) {
                raw.selectedPlateTemplate = List<string>(model.availablePlateTemplates).get(0);
            }
        }

        return new AssayProtocolModel({ ...raw, name, domains });
    }

    static serialize(model: AssayProtocolModel): any {
        // need to serialize the DomainDesign objects to remove the unrecognized fields
        const domains = model.domains.map(domain => {
            return DomainDesign.serialize(domain);
        });

        const json = model.merge({ domains }).toJS();

        // only need to serialize the id and not the autoCopyTargetContainer object
        delete json.autoCopyTargetContainer;
        delete json.exception;
        return json;
    }

    isActive(): boolean {
        return this.status === Status.Active;
    }

    getDomainByNameSuffix(name: string): DomainDesign {
        if (this.domains.size > 0) {
            return this.domains.find(domain => {
                return domain.isNameSuffixMatch(name);
            });
        }
    }

    get container(): string {
        const container = new Container(getServerContext().container);
        const domainContainerId = this.domainContainerId;

        return this.isNew()
            ? getAppHomeFolderPath(container)
            : domainContainerId === container.id
              ? container.path
              : domainContainerId;
    }

    get domainContainerId(): string {
        return this.getIn(['domains', 0, 'container']);
    }

    isNew(): boolean {
        return !this.protocolId;
    }

    allowPlateTemplateSelection(): boolean {
        return this.availablePlateTemplates && Utils.isArray(this.availablePlateTemplates);
    }

    allowDetectionMethodSelection(): boolean {
        return this.availableDetectionMethods && Utils.isArray(this.availableDetectionMethods);
    }

    allowMetadataInputFormatSelection(): boolean {
        return (
            this.availableMetadataInputFormats &&
            Utils.isObject(this.availableMetadataInputFormats) &&
            !Utils.isEmptyObj(this.availableMetadataInputFormats)
        );
    }

    validateTransformScripts(): string {
        if (this.protocolTransformScripts === undefined || this.protocolTransformScripts.size === 0) {
            return undefined;
        }

        // make sure we don't have any script inputs that are empty strings
        const hasEmptyScript = this.protocolTransformScripts.some(
            config => config?.scriptPath === undefined || config.scriptPath === null || config.scriptPath.length === 0
        );
        if (hasEmptyScript) {
            return 'Missing required transform script path.';
        }
    }

    getFirstDomainFieldError(): FieldErrors {
        const firstErrantDomain = this.domains.find(domain => domain.hasInvalidFields());
        return firstErrantDomain !== undefined ? firstErrantDomain.getInvalidFields().first().getErrors() : undefined;
    }

    isValid(): boolean {
        return this.hasValidProperties() && this.getFirstDomainFieldError() === undefined;
    }

    isGPAT(): boolean {
        return this.providerName.toLowerCase() === 'general';
    }

    hasValidProperties(): boolean {
        return (
            this.name !== undefined &&
            this.name !== null &&
            this.name.trim().length > 0 &&
            (!this.allowMetadataInputFormatSelection() || Utils.isString(this.selectedMetadataInputFormat)) &&
            (!this.allowDetectionMethodSelection() || Utils.isString(this.selectedDetectionMethod)) &&
            (!this.allowPlateTemplateSelection() || Utils.isString(this.selectedPlateTemplate)) &&
            this.validateTransformScripts() === undefined
        );
    }

    get hasBatchFields(): boolean {
        return this.getDomainByNameSuffix('Batch')?.fields.size > 0;
    }

    getUrl(): AppURL {
        return AppURL.create('assays', this.providerName, this.name);
    }
}
