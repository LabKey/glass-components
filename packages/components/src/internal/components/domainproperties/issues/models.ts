/*
 * Copyright (c) 2020 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { immerable } from 'immer';

import { DomainDesign } from '../models';

export interface IssuesListDefOptionsConfig {
    entityId?: string;
    issueDefName: string;
    singularItemName: string;
    pluralItemName: string;
    commentSortDirection: string;
    assignedToGroup: number;
    assignedToUser: number;
}

interface IssuesListDefModelConfig extends IssuesListDefOptionsConfig {
    exception: string;
    domain: DomainDesign;
    domainId: number;
    domainKindName: string;
}

export class IssuesListDefModel implements IssuesListDefModelConfig {
    [immerable] = true;

    readonly exception: string;
    readonly domain: DomainDesign;
    readonly entityId?: string;
    readonly domainId: number;
    readonly issueDefName: string;
    readonly singularItemName: string = 'Issue';
    readonly pluralItemName: string = 'Issues';
    readonly commentSortDirection: string = 'ASC';
    readonly assignedToGroup: number;
    readonly assignedToUser: number;
    readonly domainKindName: string;

    constructor(values?: Partial<IssuesListDefModelConfig>) {
        Object.assign(this, values);
    }

    static create(raw: any, defaultSettings = null): IssuesListDefModel {
        if (defaultSettings) {
            const domain = DomainDesign.create(undefined);
            return new IssuesListDefModel({ ...defaultSettings, domain });
        } else {
            const domain = DomainDesign.create(raw.domainDesign);
            return new IssuesListDefModel({ ...raw.options, domain });
        }
    }

    isNew(): boolean {
        return !this.entityId;
    }

    isValid(): boolean {
        return this.hasValidProperties() && !this.domain.hasInvalidFields();
    }

    hasValidProperties(): boolean {
        return this.issueDefName !== undefined && this.issueDefName !== null && this.issueDefName.trim().length > 0;
    }

    getOptions(): IssuesListDefOptionsConfig {
        return {
            entityId: this.entityId,
            issueDefName: this.issueDefName,
            singularItemName: this.singularItemName,
            pluralItemName: this.pluralItemName,
            commentSortDirection: this.commentSortDirection,
            assignedToGroup: this.assignedToGroup,
            assignedToUser: this.assignedToUser,
        };
    }
}
