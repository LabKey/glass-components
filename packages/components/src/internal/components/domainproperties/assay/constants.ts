export const PROPERTIES_HEADER_ID = 'assay-properties-hdr';
export const FORM_ID_PREFIX = 'assay-design-';

export const FORM_IDS = {
    ASSAY_NAME: FORM_ID_PREFIX + 'name',
    ASSAY_DESCRIPTION: FORM_ID_PREFIX + 'description',
    AUTO_LINK_TARGET: FORM_ID_PREFIX + 'autoCopyTargetContainerId',
    AUTO_LINK_CATEGORY: FORM_ID_PREFIX + 'autoLinkCategory',
    BACKGROUND_UPLOAD: FORM_ID_PREFIX + 'backgroundUpload',
    DETECTION_METHOD: FORM_ID_PREFIX + 'selectedDetectionMethod',
    EDITABLE_RUNS: FORM_ID_PREFIX + 'editableRuns',
    EDITABLE_RESULTS: FORM_ID_PREFIX + 'editableResults',
    METADATA_INPUT_FORMAT: FORM_ID_PREFIX + 'selectedMetadataInputFormat',
    PLATE_TEMPLATE: FORM_ID_PREFIX + 'selectedPlateTemplate',
    PROTOCOL_TRANSFORM_SCRIPTS: FORM_ID_PREFIX + 'protocolTransformScripts',
    QC_ENABLED: FORM_ID_PREFIX + 'qcEnabled',
    SAVE_SCRIPT_FILES: FORM_ID_PREFIX + 'saveScriptFiles',
    PLATE_METADATA: FORM_ID_PREFIX + 'plateMetadata',
    STATUS: FORM_ID_PREFIX + 'status',
};

export const BOOLEAN_FIELDS = [
    FORM_IDS.BACKGROUND_UPLOAD,
    FORM_IDS.EDITABLE_RUNS,
    FORM_IDS.EDITABLE_RESULTS,
    FORM_IDS.QC_ENABLED,
    FORM_IDS.SAVE_SCRIPT_FILES,
    FORM_IDS.PLATE_METADATA,
    FORM_IDS.STATUS,
];
