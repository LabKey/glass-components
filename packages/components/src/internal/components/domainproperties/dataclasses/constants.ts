export const DATACLASS_DOMAIN_SYSTEM_FIELDS = [
    {
        Name: 'Name',
        Label: 'Name',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a short name for this data object.',
    },
    {
        Name: 'Description',
        Label: 'Description',
        'Data Type': 'Text',
        Required: false,
        Description: 'Contains a description for this data object',
    },
];
