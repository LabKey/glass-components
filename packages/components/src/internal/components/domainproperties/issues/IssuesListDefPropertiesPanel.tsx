import React from 'react';

import { Col, Form, Row } from 'react-bootstrap';

import { Utils } from '@labkey/api';

import produce from 'immer';

import { BasePropertiesPanel, BasePropertiesPanelProps } from '../BasePropertiesPanel';
import {
    InjectedDomainPropertiesPanelCollapseProps,
    withDomainPropertiesPanelCollapse,
} from '../DomainPropertiesPanelCollapse';

import { HelpTopicURL } from '../HelpTopicURL';

import { DEFINE_ISSUES_LIST_TOPIC } from '../../../../util/helpLinks';

import { AssignmentOptions, BasicPropertiesFields } from './IssuesListDefPropertiesPanelFormElements';
import { IssuesListDefModel } from './models';

const PROPERTIES_HEADER_ID = 'issues-properties-hdr';

interface OwnProps {
    model: IssuesListDefModel;
    onChange: (model: IssuesListDefModel) => void;
    successBsStyle?: string;
}

type Props = OwnProps & BasePropertiesPanelProps;

interface State {
    isValid: boolean;
}

export class IssuesListDefPropertiesPanelImpl extends React.PureComponent<
    Props & InjectedDomainPropertiesPanelCollapseProps,
    State
> {
    constructor(props: Props & InjectedDomainPropertiesPanelCollapseProps) {
        super(props);

        this.state = produce(
            {
                isValid: true,
            },
            () => {}
        );
    }

    updateValidStatus = (newModel?: IssuesListDefModel) => {
        const { model, onChange } = this.props;
        const updatedModel = newModel || model;
        const isValid = updatedModel && updatedModel.hasValidProperties();
        this.setState(
            () => ({ isValid }),
            () => {
                if (newModel) {
                    onChange(updatedModel);
                }
            }
        );
    };

    onInputChange = e => {
        const id = e.target.id;
        let value = e.target.value;

        // special case for empty string, set as null instead
        if (Utils.isString(value) && value.length === 0) {
            value = null;
        }

        this.onChange(id, value);
    };

    onChange = (identifier: string, value: any, clearingField?: string): void => {
        const { model } = this.props;
        const newModel = produce(model, (draft: IssuesListDefModel) => {
            draft[identifier] = value;
            if (clearingField) {
                draft[clearingField] = undefined;
            }
        });
        this.updateValidStatus(newModel);
    };

    onSelectChange = (name, value) => {
        if (name === 'assignedToGroup') {
            this.onChange(name, value, 'assignedToUser');
        } else {
            this.onChange(name, value);
        }
    };

    render() {
        const { model } = this.props;
        const { isValid } = this.state;
        return (
            <BasePropertiesPanel
                {...this.props}
                headerId={PROPERTIES_HEADER_ID}
                title="Issues List Properties"
                titlePrefix={model.issueDefName}
                updateValidStatus={this.updateValidStatus}
                isValid={isValid}
            >
                <Row className="margin-bottom">
                    <Col xs={12}>
                        <HelpTopicURL helpTopic={DEFINE_ISSUES_LIST_TOPIC} nounPlural="issues lists" />
                    </Col>
                </Row>
                <Form>
                    <BasicPropertiesFields
                        model={model}
                        onInputChange={this.onInputChange}
                        onSelect={this.onSelectChange}
                    />
                    <AssignmentOptions model={model} onSelect={this.onSelectChange} />
                </Form>
            </BasePropertiesPanel>
        );
    }
}

export const IssuesListDefPropertiesPanel = withDomainPropertiesPanelCollapse<Props>(IssuesListDefPropertiesPanelImpl);
