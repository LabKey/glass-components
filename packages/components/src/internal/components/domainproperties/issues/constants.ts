export const ISSUES_LIST_DEF_SINGULAR_PLURAL_TIP =
    'The Singular item name and Plural item name fields control the display ' +
    'name for an "issue" across the Issues module. For example, you might instead refer to issues as "Bugs", "Tickets", or "Opportunities" depending on the context.';

export const ISSUES_LIST_DEF_SORT_DIRECTION_TIP =
    'By default, comments on an issue are shown in the order they are added, oldest first. Change the Comment Sort Direction to newest first if you prefer.';
export const ISSUES_LIST_GROUP_ASSIGN_TIP =
    "Select a specific group from the dropdown, both site and project level groups are available. If left blank, the 'Assigned To' field will display all users who are a) assigned the role of Editor or higher in the folder, and b) are members of at least one project-level security group.";
export const ISSUES_LIST_USER_ASSIGN_TIP =
    'In some workflows it is useful to have a default user to whom all issues are assigned, such as for initial triage and balancing assignments across a group. If left blank, no user will be assigned by default.';
export const ISSUES_LIST_RELATED_FOLDER_TIP =
    'A related issue can be created in any issue list that the user can insert into. This setting controls the default folder selection in the related issue dialog box.';
export const ISSUES_LIST_RESTRICTED_TRACKER_TIP =
    'A restricted issue list will limit viewing and updating of all issues in the list. When the checkbox is selected only users who are either assigned to the issue or on the notify list ' +
    'will have access to an issue. An additional group can optionally be configured to provide access to any members of that group.';
export const ISSUES_LIST_RESTRICTED_GROUP_TIP =
    'An optional project or site group that can give members of the group access to the issues in a restricted issue list.';
