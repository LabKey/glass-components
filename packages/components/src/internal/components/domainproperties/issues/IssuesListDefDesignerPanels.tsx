import React from 'react';
import { List } from 'immutable';
import produce, { Draft } from 'immer';

import { BaseDomainDesigner, InjectedBaseDomainDesignerProps, withBaseDomainDesigner } from '../BaseDomainDesigner';
import { getDomainPanelStatus, saveDomain } from '../actions';
import { resolveErrorMessage } from '../../../util/messaging';
import DomainForm from '../DomainForm';
import { DomainDesign } from '../models';

import { IssuesListDefPropertiesPanel } from './IssuesListDefPropertiesPanel';
import { IssuesListDefModel } from './models';
import { saveIssueListDefOptions } from './actions';

interface Props {
    initModel?: IssuesListDefModel;
    onChange?: (model: IssuesListDefModel) => void;
    onCancel: () => void;
    onComplete: (model: IssuesListDefModel) => void;
    useTheme?: boolean;
    containerTop?: number; // This sets the top of the sticky header, default is 0
    successBsStyle?: string;
    saveBtnText?: string;
}

interface State {
    model: IssuesListDefModel;
}

class IssuesDesignerPanelsImpl extends React.PureComponent<Props & InjectedBaseDomainDesignerProps, State> {
    constructor(props: Props & InjectedBaseDomainDesignerProps) {
        super(props);
        this.state = produce(
            {
                model: props.initModel || IssuesListDefModel.create({}),
            },
            () => {}
        );
    }

    onPropertiesChange = (model: IssuesListDefModel) => {
        const { onChange } = this.props;

        this.setState(
            produce((draft: Draft<State>) => {
                draft.model = model;
            }),
            () => {
                if (onChange) {
                    onChange(model);
                }
            }
        );
    };

    onDomainChange = (domain: DomainDesign, dirty: boolean) => {
        const { onChange } = this.props;

        this.setState(
            produce((draft: Draft<State>) => {
                draft.model.domain = domain;
            }),
            () => {
                if (onChange && dirty) {
                    onChange(this.state.model);
                }
            }
        );
    };

    onFinish = () => {
        const { model } = this.state;
        this.props.onFinish(model.isValid(), model.domain.isSharedDomain() ? this.saveOptions : this.saveDomain);
    };

    saveOptions = () => {
        const { setSubmitting } = this.props;
        const { model } = this.state;

        saveIssueListDefOptions(model.getOptions())
            .then(() => this.onSaveComplete())
            .catch(response => {
                setSubmitting(false, () => {
                    this.setState(
                        produce((draft: Draft<State>) => {
                            draft.model.exception = response.exception;
                        })
                    );
                });
            });
    };

    saveDomain = () => {
        const { setSubmitting } = this.props;
        const { model } = this.state;

        saveDomain(model.domain, model.domainKindName, model.getOptions(), model.issueDefName)
            .then(response => this.onSaveComplete(response))
            .catch(response => {
                const exception = resolveErrorMessage(response);

                setSubmitting(false, () => {
                    this.setState(
                        produce((draft: Draft<State>) => {
                            if (exception) {
                                draft.model.exception = exception;
                            } else {
                                draft.model.exception = undefined;
                                draft.model.domain = response;
                            }
                        })
                    );
                });
            });
    };

    onSaveComplete = (response?: any) => {
        const { setSubmitting } = this.props;

        this.setState(
            produce((draft: Draft<State>) => {
                draft.model.exception = undefined;
                if (response) {
                    draft.model.domain = response;
                }
            }),
            () => {
                setSubmitting(false, () => {
                    const { model } = this.state;
                    this.props.onComplete(model);
                });
            }
        );
    };

    render() {
        const {
            onCancel,
            useTheme,
            containerTop,
            successBsStyle,
            visitedPanels,
            currentPanelIndex,
            firstState,
            validatePanel,
            submitting,
            onTogglePanel,
            saveBtnText,
        } = this.props;
        const { model } = this.state;

        return (
            <BaseDomainDesigner
                name={model.issueDefName}
                exception={model.exception}
                domains={List.of(model.domain)}
                hasValidProperties={model.hasValidProperties()}
                visitedPanels={visitedPanels}
                submitting={submitting}
                onCancel={onCancel}
                onFinish={this.onFinish}
                saveBtnText={saveBtnText}
                successBsStyle={successBsStyle}
            >
                <IssuesListDefPropertiesPanel
                    model={model}
                    onChange={this.onPropertiesChange}
                    controlledCollapse={true}
                    initCollapsed={currentPanelIndex !== 0}
                    panelStatus={
                        model.isNew()
                            ? getDomainPanelStatus(0, currentPanelIndex, visitedPanels, firstState)
                            : 'COMPLETE'
                    }
                    validate={validatePanel === 0}
                    onToggle={(collapsed, callback) => {
                        onTogglePanel(0, collapsed, callback);
                    }}
                    useTheme={useTheme}
                    successBsStyle={successBsStyle}
                />
                <DomainForm
                    key={model.domain.domainId || 0}
                    domainIndex={0}
                    domain={model.domain}
                    headerTitle="Fields"
                    helpNoun="issues list"
                    helpTopic={null} // null so that we don't show the "learn more about this tool" link for this domains
                    onChange={this.onDomainChange}
                    controlledCollapse={true}
                    initCollapsed={currentPanelIndex !== 1}
                    validate={validatePanel === 1}
                    panelStatus={
                        model.isNew()
                            ? getDomainPanelStatus(1, currentPanelIndex, visitedPanels, firstState)
                            : 'COMPLETE'
                    }
                    containerTop={containerTop}
                    onToggle={(collapsed, callback) => {
                        onTogglePanel(1, collapsed, callback);
                    }}
                    useTheme={useTheme}
                    successBsStyle={successBsStyle}
                    domainFormDisplayOptions={{
                        isDragDisabled: model.domain.isSharedDomain(),
                        hideAddFieldsButton: model.domain.isSharedDomain(),
                    }}
                />
            </BaseDomainDesigner>
        );
    }
}

export const IssuesListDefDesignerPanels = withBaseDomainDesigner<Props>(IssuesDesignerPanelsImpl);
