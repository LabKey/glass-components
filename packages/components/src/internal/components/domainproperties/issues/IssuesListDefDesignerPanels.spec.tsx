import React from 'react';
import renderer from 'react-test-renderer';

import { mount } from 'enzyme';

import DomainForm from '../DomainForm';

import { PROPERTIES_PANEL_ERROR_MSG } from '../constants';

import { initUnitTestMocks } from '../../../../test/testHelperMocks';

import { Alert } from '../../base/Alert';

import { IssuesListDefPropertiesPanel } from './IssuesListDefPropertiesPanel';
import { IssuesListDefDesignerPanels } from './IssuesListDefDesignerPanels';
import { IssuesListDefModel } from './models';

const emptyNewModel = IssuesListDefModel.create(null, { issueDefName: 'Issues List For Jest' });

const BASE_PROPS = {
    onComplete: jest.fn(),
    onCancel: jest.fn(),
    testMode: true,
};

beforeAll(() => {
    initUnitTestMocks();
});

describe('IssuesListDefDesignerPanel', () => {
    test('new Issue List Definition', () => {
        const issuesDesignerPanels = <IssuesListDefDesignerPanels {...BASE_PROPS} initModel={emptyNewModel} />;

        const tree = renderer.create(issuesDesignerPanels);
        expect(tree).toMatchSnapshot();
    });

    test('visible properties', () => {
        const issuesDesignerPanels = mount(<IssuesListDefDesignerPanels {...BASE_PROPS} initModel={emptyNewModel} />);

        expect(issuesDesignerPanels.find(IssuesListDefPropertiesPanel)).toHaveLength(1);
        expect(issuesDesignerPanels.find(DomainForm)).toHaveLength(1);
        issuesDesignerPanels.unmount();
    });

    test('open fields panel', () => {
        const wrapped = mount(<IssuesListDefDesignerPanels {...BASE_PROPS} />);

        const panelHeader = wrapped.find('div#domain-header');
        expect(wrapped.find('#domain-header').at(2).hasClass('domain-panel-header-collapsed')).toBeTruthy();
        panelHeader.simulate('click');
        expect(wrapped.find('#domain-header').at(2).hasClass('domain-panel-header-expanded')).toBeTruthy();

        expect(wrapped.find(Alert)).toHaveLength(2);
        expect(wrapped.find(Alert).at(0).text()).toEqual(PROPERTIES_PANEL_ERROR_MSG);
        expect(wrapped.find(Alert).at(1).text()).toEqual(
            'Please correct errors in the properties panel before saving.'
        );
        wrapped.unmount();
    });

    // TODO: Add test cases for editing Issues List Def in the near future
});
