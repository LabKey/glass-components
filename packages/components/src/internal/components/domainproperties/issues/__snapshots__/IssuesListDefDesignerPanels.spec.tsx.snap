// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IssuesListDefDesignerPanel new Issue List Definition 1`] = `
<BaseDomainDesigner
  domains={
    Immutable.List [
      Immutable.Record {
        "name": undefined,
        "container": undefined,
        "description": undefined,
        "domainURI": undefined,
        "domainId": null,
        "allowFileLinkProperties": false,
        "allowAttachmentProperties": false,
        "allowFlagProperties": true,
        "allowSampleSubjectProperties": true,
        "allowTextChoiceProperties": true,
        "allowTimepointProperties": false,
        "showDefaultValueSettings": false,
        "defaultDefaultValueType": undefined,
        "defaultValueOptions": Immutable.List [],
        "fields": Immutable.List [],
        "indices": Immutable.List [],
        "domainException": undefined,
        "mandatoryFieldNames": Immutable.List [],
        "reservedFieldNames": Immutable.List [],
        "newDesignFields": undefined,
        "instructions": undefined,
        "domainKindName": undefined,
        "schemaName": undefined,
        "queryName": undefined,
        "disabledSystemFields": undefined,
      },
    ]
  }
  hasValidProperties={true}
  name="Issues List For Jest"
  onCancel={[MockFunction]}
  onFinish={[Function]}
  saveBtnText="Save"
  submitting={false}
  visitedPanels={Immutable.List []}
>
  <ComponentWithDomainPropertiesPanelCollapse
    controlledCollapse={true}
    initCollapsed={false}
    model={
      IssuesListDefModel {
        "commentSortDirection": "ASC",
        "domain": Immutable.Record {
          "name": undefined,
          "container": undefined,
          "description": undefined,
          "domainURI": undefined,
          "domainId": null,
          "allowFileLinkProperties": false,
          "allowAttachmentProperties": false,
          "allowFlagProperties": true,
          "allowSampleSubjectProperties": true,
          "allowTextChoiceProperties": true,
          "allowTimepointProperties": false,
          "showDefaultValueSettings": false,
          "defaultDefaultValueType": undefined,
          "defaultValueOptions": Immutable.List [],
          "fields": Immutable.List [],
          "indices": Immutable.List [],
          "domainException": undefined,
          "mandatoryFieldNames": Immutable.List [],
          "reservedFieldNames": Immutable.List [],
          "newDesignFields": undefined,
          "instructions": undefined,
          "domainKindName": undefined,
          "schemaName": undefined,
          "queryName": undefined,
          "disabledSystemFields": undefined,
        },
        "issueDefName": "Issues List For Jest",
        "pluralItemName": "Issues",
        "singularItemName": "Issue",
        Symbol(immer-draftable): true,
      }
    }
    onChange={[Function]}
    onToggle={[Function]}
    panelStatus="NONE"
    validate={true}
  />
  <DomainForm
    controlledCollapse={true}
    domain={
      Immutable.Record {
        "name": undefined,
        "container": undefined,
        "description": undefined,
        "domainURI": undefined,
        "domainId": null,
        "allowFileLinkProperties": false,
        "allowAttachmentProperties": false,
        "allowFlagProperties": true,
        "allowSampleSubjectProperties": true,
        "allowTextChoiceProperties": true,
        "allowTimepointProperties": false,
        "showDefaultValueSettings": false,
        "defaultDefaultValueType": undefined,
        "defaultValueOptions": Immutable.List [],
        "fields": Immutable.List [],
        "indices": Immutable.List [],
        "domainException": undefined,
        "mandatoryFieldNames": Immutable.List [],
        "reservedFieldNames": Immutable.List [],
        "newDesignFields": undefined,
        "instructions": undefined,
        "domainKindName": undefined,
        "schemaName": undefined,
        "queryName": undefined,
        "disabledSystemFields": undefined,
      }
    }
    domainFormDisplayOptions={
      {
        "hideAddFieldsButton": false,
        "hideImportExport": true,
        "hideInferFromFile": true,
        "isDragDisabled": false,
      }
    }
    domainIndex={0}
    headerTitle="Fields"
    helpNoun="issues list"
    helpTopic={null}
    initCollapsed={true}
    key="0"
    onChange={[Function]}
    onToggle={[Function]}
    panelStatus="TODO"
    testMode={true}
    validate={false}
  />
</BaseDomainDesigner>
`;
